### PR TITLE
Reimplement silhouette in a mem constant way, pdist using lax scan

### LINKED ISSUE #36: Memory issue for ASW scores
Hi,

thanks for the cool reimplementation if the scib metrics!
I've tried out the ASW metrics (label and batch) on a 700MB anndata object but get out-of-memory errors even at 10GB. Is there potentially a memory leak?

Best,
Michaela

### PR DESCRIPTION
**Pull request by adamgayoso**
Fixes #36 

Our implementation now exactly follows the algorithm used in sklearn (instead of computing full X pairwise distance matrix, compute it in vertical chunks `cdist(X_chunk_size, X)` and then aggregate intra and inter clust dists

### CODE REVIEW

File: src/scib_metrics/utils/_silhouette.py [hunk 1]
```diff
@@ -131,39 +84,24 @@ def silhouette_samples(X: np.ndarray, labels: np.ndarray) -> np.ndarray:
...
+    label_freqs = jnp.bincount(labels)
+    reduce_fn = partial(_silhouette_reduce, labels=labels, label_freqs=label_freqs)
+    results = _pairwise_distances_chunked(X, chunk_size=chunk_size, reduce_fn=reduce_fn)
+    intra_clust_dists, inter_clust_dists = results
+
+    denom = jnp.take(label_freqs - 1, labels, mode="clip")
+    intra_clust_dists /= denom
+    sil_samples = inter_clust_dists - intra_clust_dists
+    sil_samples /= np.maximum(intra_clust_dists, inter_clust_dists)
+    return np.array(jax.device_get(jnp.nan_to_num(sil_samples)))
```
Comment (martinkim0) [inline src/scib_metrics/utils/_silhouette.py:107]:
```suggestion
    return get_ndarray(jnp.nan_to_num(sil_samples))
```
from `utils`.

File: src/scib_metrics/utils/_silhouette.py [hunk 2]
```diff
@@ -131,39 +84,24 @@ def silhouette_samples(X: np.ndarray, labels: np.ndarray) -> np.ndarray:
...
+    labels = jnp.asarray(labels)
+    label_freqs = jnp.bincount(labels)
+    reduce_fn = partial(_silhouette_reduce, labels=labels, label_freqs=label_freqs)
+    results = _pairwise_distances_chunked(X, chunk_size=chunk_size, reduce_fn=reduce_fn)
+    intra_clust_dists, inter_clust_dists = results
+
+    denom = jnp.take(label_freqs - 1, labels, mode="clip")
+    intra_clust_dists /= denom
+    sil_samples = inter_clust_dists - intra_clust_dists
+    sil_samples /= np.maximum(intra_clust_dists, inter_clust_dists)
```
Comment (martinkim0) [inline src/scib_metrics/utils/_silhouette.py:106]: Is there a reason why `jnp` wasn't used here?

File: src/scib_metrics/utils/_silhouette.py [hunk 3]
```diff
@@ -1,126 +1,79 @@
...
 from typing import Tuple, Union
 
-import chex
 import jax
 import jax.numpy as jnp
 import numpy as np
+import pandas as pd
 
-from ._dist import cdist, pdist_squareform
+from ._dist import cdist
```
Comment (martinkim0) [inline src/scib_metrics/utils/_silhouette.py:9]:
```suggestion
from ._types import NdArray
```

File: src/scib_metrics/utils/_silhouette.py [hunk 4]
```diff
@@ -1,126 +1,79 @@
...
 
 NdArray = Union[np.ndarray, jnp.ndarray]
```
Comment (martinkim0) [inline src/scib_metrics/utils/_silhouette.py:11]:
```suggestion
```

**Review by martinkim0 — APPROVED** [review]
